# Tasks 20–22: Group-Level Survival Analysis

In [1]:
import numpy as np
import pandas as pd

df = pd.read_csv("../Data/Titanic-Dataset.csv")
df = df.drop_duplicates().copy()
df["Age"] = df["Age"].fillna(df["Age"].median())
df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])

In [2]:
# Task 20: Overall survival rate
survival_rate = df["Survived"].mean() * 100
print(f"Overall Survival Rate: {survival_rate:.2f}%")

Overall Survival Rate: 38.38%


**Student Task (Task 20):**
Overall survival rate = 38.38%

In [3]:
# Task 21: Survival rate by sex
df.groupby("Sex")["Survived"].mean() * 100

Sex
female    74.203822
male      18.890815
Name: Survived, dtype: float64

**Student Questions (Task 21):**
1. Survival rate for females: 74.20%
2. Survival rate for males: 18.89%
3. Females had the higher survival rate
4. The difference is very large (over 55 percentage points) — consistent with the "women and children first" evacuation policy on the Titanic.

In [4]:
# Task 22: Survival rate by passenger class
df.groupby("Pclass")["Survived"].mean() * 100

Pclass
1    62.962963
2    47.282609
3    24.236253
Name: Survived, dtype: float64

**Student Questions (Task 22):**
1st Class = 62.96%
2nd Class = 47.28%
3rd Class = 24.24%

Which passenger class had the highest survival rate? 1st class — survival rate dropped 
steadily as class declined, suggesting wealthier/higher-class passengers had better access to lifeboats.